libcudacxx/include/cuda/std/__random/linear_congruential_engine.h

```diff
@@ -236,108 +238,151 @@ class _CCCL_TYPE_VISIBILITY_DEFAULT linear_congruential_engine
   static constexpr const result_type default_seed = 1u;
 
   // constructors and seeding functions
-  _CCCL_API linear_congruential_engine() noexcept
+  _CCCL_API constexpr linear_congruential_engine() noexcept
       : linear_congruential_engine(default_seed)
   {}
-  _CCCL_API explicit linear_congruential_engine(result_type __s) noexcept
+  _CCCL_API explicit constexpr linear_congruential_engine(result_type __s) noexcept
   {
     seed(__s);
   }
 
   template <class _Sseq, enable_if_t<__is_seed_sequence<_Sseq, linear_congruential_engine>, int> = 0>
-  _CCCL_API explicit linear_congruential_engine(_Sseq& __q) noexcept
+  _CCCL_API explicit constexpr linear_congruential_engine(_Sseq& __q) noexcept
   {
     seed(__q);
   }
-  _CCCL_API void seed(result_type __s = default_seed)
+  _CCCL_API constexpr void seed(result_type __s = default_seed) noexcept
   {
     seed(integral_constant<bool, __m == 0>(), integral_constant<bool, __c == 0>(), __s);
   }
   template <class _Sseq, enable_if_t<__is_seed_sequence<_Sseq, linear_congruential_engine>, int> = 0>
-  _CCCL_API void seed(_Sseq& __q) noexcept
+  _CCCL_API constexpr void seed(_Sseq& __q) noexcept
   {
     __seed(__q,
            integral_constant<uint32_t,
                              1 + (__m == 0 ? (sizeof(result_type) * CHAR_BIT - 1) / 32 : (__m > 0x100000000ull))>());
   }
 
   // generating functions
-  [[nodiscard]] _CCCL_API result_type operator()() noexcept
+  _CCCL_API constexpr result_type operator()() noexcept
   {
     return __x_ = static_cast<result_type>(__lce_ta<__a, __c, __m, _Mp>::next(__x_));
   }
-  _CCCL_API void discard(uint64_t __z) noexcept
+
+  _CCCL_API constexpr void discard(uint64_t __z) noexcept
   {
-    for (; __z; --__z)
+    constexpr bool __can_overflow = (__a != 0 && __m != 0 && __m - 1 > (_Mp - __c) / __a);
+    // Fallback implementation
+    if constexpr (__can_overflow)
+    {
+      for (; __z; --__z)
+      {
+        (void) operator()();
+      }
+    }
+    else if constexpr (increment == 0)
     {
-      (void) operator()();
+      // minstd_rand and minstd_rand0 can use this efficient implementation
+      uint64_t __acc_mult = 1;
+      uint64_t __cur_mult = multiplier;
+      while (__z > 0)
+      {
+        if (__z & 1)
+        {
+          __acc_mult = (__acc_mult * __cur_mult) % modulus;
```

Comment:
I would prefer if we had this as
```suggestion
          __acc_mult = (__acc_mult * __cur_mult) % modulus;
          if constexpr(increment != 0) {
            __acc_plus = (__acc_plus * __cur_mult + __cur_plus) % modulus;          
          }
```

We would need to annotate `__acc_plus ` with `[[maybe_unused]]` but that seems to be much cleaner

Reply:
Yeah this is nicer.